network/state/switcher.go

```diff
@@ -17,19 +17,46 @@
 package state
 
 import (
+	"context"
+	"sync"
+
 	"github.com/insolar/insolar/core"
+	"github.com/insolar/insolar/instrumentation/inslogger"
 )
 
 // NetworkSwitcher is a network FSM using for bootstrapping
 type NetworkSwitcher struct {
+	NodeNetwork        core.NodeNetwork        `inject:""`
+	Certificate        core.Certificate        `inject:""`
+	SwitcherWorkAround core.SwitcherWorkAround `inject:""`
+
+	state     core.NetworkState
+	stateLock *sync.RWMutex
 }
 
 // NewNetworkSwitcher creates new NetworkSwitcher
 func NewNetworkSwitcher() (*NetworkSwitcher, error) {
-	return &NetworkSwitcher{}, nil
+	return &NetworkSwitcher{state: core.NoNetworkState}, nil
 }
 
 // GetState method returns current network state
 func (ns *NetworkSwitcher) GetState() core.NetworkState {
-	return core.CompleteNetworkState
+	ns.stateLock.Lock()
+	defer ns.stateLock.RUnlock()
+
+	return ns.state
+}
+
+// OnPulse method checks current state and finds out reasons to update this state
+func (ns *NetworkSwitcher) OnPulse(ctx context.Context, pulse core.Pulse) error {
+	inslogger.FromContext(ctx).Info("Current NetworkSwitcher state is: %s", ns.state)
+
+	if ns.SwitcherWorkAround.IsBootstrapped() {
+		ns.stateLock.Lock()
+		ns.state = core.CompleteNetworkState
+		ns.stateLock.Unlock()
+		inslogger.FromContext(ctx).Info("Current NetworkSwitcher state switched to: %s", ns.state)
```

Comment:
Concurrent access to ns.state
Save ns.state to another variable inside Lock and then use it here

Reply:
Or even print it inside lock